### PR TITLE
[shape_infer]shape inference implement of Select Detectionoutput and Shufflechannels OPs

### PR DESCRIPTION
**Pull request by luweizhou2016**
### Details:
 - *Implement shape infer for Select DetectionOutput ShuffleChannels OPs*
 - *This PR rebased the new master and merged pervious 3 PRs into one*

### Tickets:
 - *68546,68548, 68538*

### CODE REVIEW

File: ngraph/core/shape_inference/include/shuffle_channels_shape_inference.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+namespace ov {
+namespace op {
+namespace v0 {
+
+template <class T>
+void shape_infer(const ShuffleChannels* op, const std::vector<T>& input_shapes, std::vector<T>& output_shapes) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 1 && output_shapes.size() == 1);
+    const auto& input_shape = input_shapes[0];
+    NODE_VALIDATION_CHECK(op, op->m_group >= 1, "The 'group' parameter must be greater or equal to 1.");
+    if (input_shape.is_static()) {
```
Comment (jane-intel) [inline ngraph/core/shape_inference/include/shuffle_channels_shape_inference.hpp:18]: We only need input shape to have static rank and shape[axis] to be static to perform the checks

File: ngraph/core/shape_inference/include/select_shape_inference.hpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        NODE_VALIDATION_CHECK(op,
+                              then_else_merged,
+                              "'Else' tensor and `Then` tensor are not broadcasted to each other ");
+        NODE_VALIDATION_CHECK(op,
+                              T::broadcast_merge_into(result_shape, input_shapes[0], op->get_auto_broadcast()),
+                              "'Cond' tensor shape is not broadcastable.");
+    } else {
+        result_shape = input_shapes[2];
+        for (int i = 1; i >= 0; i--) {
+            if (op->get_auto_broadcast().m_type == op::AutoBroadcastType::NONE) {
```
Comment (jane-intel) [inline ngraph/core/shape_inference/include/select_shape_inference.hpp:59]: Let's have broadcast saved once somewhere?

File: ngraph/core/shape_inference/include/detection_output_shape_inference.hpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    const int num_loc_classes = op->m_attrs.share_location ? 1 : op->m_attrs.num_classes;
+    const int prior_box_size = op->m_attrs.normalized ? 4 : 5;
+
+    dim_t num_images{};
+    dim_t num_prior_boxes{};
+
+    if (box_logits_pshape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op,
+                              box_logits_pshape.size() == 2,
+                              "Box logits rank must be 2. Got " + std::to_string(box_logits_pshape.size()));
```
Comment (jane-intel) [inline ngraph/core/shape_inference/include/detection_output_shape_inference.hpp:34]: You can do it as simple as
```suggestion
                              "Box logits rank must be 2. Got ", box_logits_pshape.size());
```

File: ngraph/core/shape_inference/include/detection_output_shape_inference.hpp [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        if (num_images.is_static() && proposals_pshape[0].is_static()) {
+            int64_t proposals_1st_dim = proposals_pshape[0].get_length();
+            int64_t num_images_val = num_images.get_length();
+            NODE_VALIDATION_CHECK(op,
+                                  proposals_1st_dim == 1 || proposals_1st_dim == num_images_val,
+                                  "Proposals' first dimension is must be equal to either batch size (" +
+                                      std::to_string(num_images_val) +
+                                      ") or 1. Got: " + std::to_string(proposals_1st_dim) + ".");
+        }
+
```
Comment (jane-intel) [inline ngraph/core/shape_inference/include/detection_output_shape_inference.hpp:99]: We do not need them to be static to check equality
Please correct correspondingly down below
```suggestion
            NODE_VALIDATION_CHECK(op,
                                  proposals_pshape[0].compatible(1) || proposals_pshape[0].compatible(num_images),
                                  "Proposals' first dimension is must be equal to either batch size (", num_images,
                                  ") or 1. Got: ", proposals_pshape[0], ".");
```

File: ngraph/core/shape_inference/include/select_shape_inference.hpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        NODE_VALIDATION_CHECK(op,
+                              then_else_merged,
+                              "'Else' tensor and `Then` tensor are not broadcasted to each other ");
+        NODE_VALIDATION_CHECK(op,
+                              T::broadcast_merge_into(result_shape, input_shapes[0], op->get_auto_broadcast()),
+                              "'Cond' tensor shape is not broadcastable.");
+    } else {
+        result_shape = input_shapes[2];
+        for (int i = 1; i >= 0; i--) {
```
Comment (jane-intel) [inline ngraph/core/shape_inference/include/select_shape_inference.hpp:58]: let's be explicit with input ports instead of this ugly for

**Review by jane-intel — CHANGES_REQUESTED** [review]

**Comment by luweizhou2016** [comment]
@jane-intel , Resolved and Update in the new commit.Thx!

**Comment by luweizhou2016** [comment]
@jane-intel , Any comments on this PR?

File: src/core/shape_inference/include/select_shape_inference.hpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            result_shape = input_shapes[src_idx];
+            then_else_merged = T::broadcast_merge_into(result_shape, input_shapes[dest_idx], broadcast_spec);
+        }
+
+        NODE_VALIDATION_CHECK(op,
+                              then_else_merged,
+                              "'Else' tensor and `Then` tensor are not broadcasted to each other ");
+        NODE_VALIDATION_CHECK(op,
+                              T::broadcast_merge_into(result_shape, input_shapes[0], broadcast_spec),
+                              "'Cond' tensor shape is not broadcastable.");
```
Comment (v-Golubev) [inline src/core/shape_inference/include/select_shape_inference.hpp:56]: As I can see, broadcasting behavior is changed in comparison with old one for PDPD broadcasting type. The OV documentation also does not define such a way to handle the case when we have PDPD broadcasting.
@luweizhou2016 could you please provide PDPD documentation for Select node where I can find the rules by which the broadcasting is carried out?

File: src/core/tests/type_prop/shuffle_channels.cpp [hunk 7]
```diff
@@ -121,11 +121,29 @@ TEST(type_prop, shuffle_channels_axis_validation) {
...
+    }
+    {
+        const auto data_input_shape = Shape{8, 3, 5, 7};
+        const auto data = make_shared<op::Parameter>(element::f64, data_input_shape);
 
-    EXPECT_EQ(shuffle_channels->get_zero_based_axis(), 1);
+        const auto shuffle_channels = make_shared<op::v0::ShuffleChannels>(data, -4, group);
+        EXPECT_EQ(shuffle_channels->get_output_partial_shape(0), data_input_shape);
+    }
 }
```
Comment (v-Golubev) [inline src/core/tests/type_prop/shuffle_channels.cpp:147]: Why did you change the existing test case? Can we create another one?

File: src/core/shape_inference/include/select_shape_inference.hpp (hunk 6 above)
Comment (luweizhou2016) [reply src/core/shape_inference/include/select_shape_inference.hpp:56]: Please help to check the spec about PDPD broadcast rule in link below. Also in the PartialShape.broadcast_merge_into() implement, for PDPD src shape is expeced to broadcast to des one way in my understanding.
https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/broadcast_rules.md

**Comment by luweizhou2016** [comment]
For the friend function declaration, we have a discussion with @jane-intel before. The original intension of friend function declaration was to  update some Ngraph OP member in the shape_infer() helper. However, it causes some thread-safe issue when enabling in CPU-plugin test. So the current conclusion is likely to be okay for not using friend function.  Also we meet some compiling problems when declaring  default constant_data to be empty. So I perfer to using public interface to get varaible. Let me know whehter this makes sense to you. Thx!

**Comment by v-Golubev** [comment]
@luweizhou2016  At the moment, when we have shape_infer functions with const Op parameter, we really don't update any class fields in these functions.  And you are right, we can not use friend function in case when we can get constant reference on the class field via existing public interface. But if we don't have this getter, we should make shape_infer function a friend for this class.
As for compilation problems in the case when our function must contain a data parameter, we should consider such cases separately. Thanks for the information, I will pay attention to it during the review.
I will double-check my comments about friend functions for compliance with the above theses. If you have any further questions, please let me know. Thank you!

File: src/core/shape_inference/include/select_shape_inference.hpp (hunk 6 above)
Comment (v-Golubev) [reply src/core/shape_inference/include/select_shape_inference.hpp:56]: Yes, I agree with your changes in the `PartialShape` and `StaticShape` classes. But I would like to clarify one more point.
In the [current implementation](https://github.com/openvinotoolkit/openvino/blob/master/src/core/src/op/select.cpp#L47) we are doing broadcasting in order `then->else->cond`. In this code you change the order of the broadcast in some cases(because otherwise it will not work). And my question is does the PDPD documentation define this behavior (changing the broadcasting order) for the Select node or not?
If not, shouldn't we keep the current behavior and throw an exception in the case where broadcasting in order `then->else->cond` cannot be done? Thanks!

File: src/core/shape_inference/include/detection_output_shape_inference.hpp [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    using dim_t = typename std::iterator_traits<typename T::iterator>::value_type;
+    NODE_VALIDATION_CHECK(op, (input_shapes.size() == 3 || input_shapes.size() == 5) && output_shapes.size() == 1);
+
+    auto& ret_output_shape = output_shapes[0];
+    ret_output_shape.resize(4);
+
+    const auto& box_logits_pshape = input_shapes[0];
+    const auto& class_preds_pshape = input_shapes[1];
+    const auto& proposals_pshape = input_shapes[2];
+    const auto attrs = op->get_attrs();
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply src/core/shape_inference/include/detection_output_shape_inference.hpp:24]: Suggestion is corrected: since we can access to the constant reference of `m_attrs` through a getter, there is no need to make the function a friend. Please, make `attrs` a const reference

File: src/core/shape_inference/include/select_shape_inference.hpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace ov {
+namespace op {
+namespace v1 {
+
+template <class T>
+void shape_infer(const Select* op, const std::vector<T>& input_shapes, std::vector<T>& output_shapes) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 3 && output_shapes.size() == 1);
+
+    const auto& broadcast_spec = op->get_auto_broadcast();
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply src/core/shape_inference/include/select_shape_inference.hpp:17]: There is no point in the proposed change: conversation resolved

File: src/core/shape_inference/include/shuffle_channels_shape_inference.hpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace ov {
+namespace op {
+namespace v0 {
+
+template <class T>
+void shape_infer(const ShuffleChannels* op, const std::vector<T>& input_shapes, std::vector<T>& output_shapes) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 1 && output_shapes.size() == 1);
+
+    const auto group = op->get_group();
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply src/core/shape_inference/include/shuffle_channels_shape_inference.hpp:17]: To my mind, we still need to do it to avoid copying

File: src/core/shape_inference/include/shuffle_channels_shape_inference.hpp [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    NODE_VALIDATION_CHECK(op, group >= 1, "The 'group' parameter must be greater or equal to 1.");
+
+    const auto& input_shape = input_shapes[0];
+    const auto input_shape_rank = input_shape.rank();
+
+    if (input_shape_rank.is_static()) {
+        const auto input_rank_value = input_shape.size();
+        NODE_VALIDATION_CHECK(op, input_rank_value >= 1, "The input tensor's shape is expected to be at least 1D.");
+
+        const auto axis = op->get_axis();
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply src/core/shape_inference/include/shuffle_channels_shape_inference.hpp:27]: To my mind, we still need to do it to avoid copying

File: src/core/shape_inference/include/select_shape_inference.hpp (hunk 6 above)
Comment (luweizhou2016) [reply src/core/shape_inference/include/select_shape_inference.hpp:56]: Agree. I have just checked PDPD frontend code. When PDPD frontend construct NGrpaph Select OP, only Numpy broadcast rule is used. So I will leave the code as origininal logic.  Since the  original code logic would have confilcts with the Select OP spec description(then and else broacast to each other), just leave one line comments in the code. Also update some new test cases. Thx!

**Comment by luweizhou2016** [comment]
@v-Golubev 2nd reveiw comments applied. Pls help to reivew. Thx!

**Comment by v-Golubev** [comment]
@luweizhou2016 Please take a look on comments. Still, I would like to clarify how the Select node is arranged in the PDPD itself. Could you tell me how broadcasting works there?

File: src/core/shape_inference/include/detection_output_shape_inference.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (luweizhou2016) [reply src/core/shape_inference/include/detection_output_shape_inference.hpp]: already updated.

**Comment by luweizhou2016** [comment]
I have just noticed that I missed some comments of last review. Sorry for that. Already update.
About the Select OP using PDPD broadcast rule:
1. By now, OV PDPD frontend doesn't support all the PDPD *select OPs. The 
paddle.index_select/paddle.masked_select OPs are similiar with pytorch torch.index_select/torch.masked_select.
Also in descriptrion  there is no broadcast rules, becuase PDPD and Pytorch doen't need to outputshape to be same with one of input. 

https://pytorch.org/docs/stable/generated/torch.masked_select.html?highlight=masked_sel#torch.masked_select
https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/masked_select_cn.html#masked-select
2.  PDPD frontend did construct Select OP to support other PDPD OPs like in embedding.cpp and others, but using the NUMPY broadcast rules. so it should be broadcast to each other.  

Does above answer you questions. Thx!

**Review by v-Golubev — APPROVED** [review]

**Comment by luweizhou2016** [comment]
@v-Golubev . For the detection_output OP, merged with the master DetectionOutput V8 OP supported.  Can you pls help to reivew this OP again? Thx!

**Review by v-Golubev — APPROVED** [review]

**Comment by v-Golubev** [comment]
Please resolve merge conflicts and fix failed tests

File: src/core/src/op/util/detection_output_base.cpp [hunk 12]
```diff
@@ -338,3 +73,20 @@ bool ov::op::util::DetectionOutputBase::visit_attributes_base(AttributeVisitor&
     visitor.on_attribute("objectness_score", attrs.objectness_score);
     return true;
 }
+
+ov::Dimension DetectionOutputBase::compute_num_classes(const AttributesBase& attrs) {
+    NODE_VALIDATION_CHECK(this,
+                          3 == get_input_size() && get_input_size() == 5,
+                          "A number of arguments must be  equal to 3 or equal to 5. Got  ",
+                          get_input_size());
```
Comment (v-Golubev) [inline src/core/src/op/util/detection_output_base.cpp:81]: It will probably fix failed tests
```suggestion
    const auto input_size = get_input_size();
    NODE_VALIDATION_CHECK(this,
                          input_size ==3 || input_size == 5,
                          "A number of arguments must be  equal to 3 or equal to 5. Got  ",
                          get_input_size());
```

Comment (luweizhou2016) [reply src/core/src/op/util/detection_output_base.cpp]: Thx for the comments. Applied.

File: src/core/shape_inference/include/detection_output_shape_inference.hpp [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+            num_prior_boxes = box_logits_pshape_2nd_dim / (num_loc_classes * 4);
+        }
+    }
+    if (class_preds_pshape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op,
+                              class_preds_pshape.size() == 2,
+                              "Class predictions rank must be 2. Got ",
+                              class_preds_pshape.size());
+        if ((!dim_num_images_updated || dim_num_images.is_dynamic()) && class_preds_pshape[0].is_static()) {
+            dim_num_images = class_preds_pshape[0].get_length();
```
Comment (jane-intel) [inline src/core/shape_inference/include/detection_output_shape_inference.hpp:178]: seems we don't need to get_length()

File: src/core/shape_inference/include/detection_output_shape_inference.hpp [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+            }
+        }
+    }
+    if (proposals_pshape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op,
+                              proposals_pshape.size() == 3,
+                              "Proposals rank must be 3. Got ",
+                              proposals_pshape.size());
+        NODE_VALIDATION_CHECK(op,
+                              proposals_pshape[0].compatible(1) || proposals_pshape[0].compatible(dim_num_images),
```
Comment (jane-intel) [inline src/core/shape_inference/include/detection_output_shape_inference.hpp:219]: nit: it's a `broadcast_merge` of dimension

File: src/core/shape_inference/include/shuffle_channels_shape_inference.hpp [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    if (input_shape_rank.is_static()) {
+        const auto input_rank_value = input_shape.size();
+        NODE_VALIDATION_CHECK(op, input_rank_value >= 1, "The input tensor's shape is expected to be at least 1D.");
+
+        const auto& axis = op->get_axis();
+        size_t axis_zb = static_cast<size_t>(axis >= 0 ? axis : (axis + static_cast<int64_t>(input_rank_value)));
+        NODE_VALIDATION_CHECK(op,
+                              axis_zb < input_rank_value,
+                              "The 'axis' parameter for ShuffleChannels has to point to one of the "
+                              "input tensor's shape dimensions.");
```
Comment (jane-intel) [inline src/core/shape_inference/include/shuffle_channels_shape_inference.hpp:32]: insufficient check. it assumes that `axis` is already in acceptable range
how about `axis = -8` and shape `[1, 2, 3]`?

File: src/core/shape_inference/include/detection_output_shape_inference.hpp [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+    val_type num_classes = 0;
+    val_type num_prior_boxes = 0;
+    dim_t dim_num_images{};
+    bool dim_num_images_updated = false;
+
+    if (dynamic_cast<const ov::op::v8::DetectionOutput*>(op)) {
+        ov::op::util::compute_num_classes(op, attrs, input_shapes, num_classes, num_prior_boxes);
+    } else {
+        num_classes = static_cast<val_type>(attribute_num_classes);
+    }
```
Comment (jane-intel) [inline src/core/shape_inference/include/detection_output_shape_inference.hpp:148]: please get rid of this dynamic cast. It seems unnecessary to me. split shape infer into two if it's needed.

**Review by jane-intel — CHANGES_REQUESTED** [review]

File: src/core/shape_inference/include/detection_output_shape_inference.hpp (hunk 14 above)
Comment (luweizhou2016) [reply src/core/shape_inference/include/detection_output_shape_inference.hpp:219]: Did you mean the broadcast_merge( result ,  proposals_pshape[0], dim_num_images)?  I think this can't be the same with the above.   Assuming proposals_pshape[0] = [1, 12], dim_num_images = [13, 20]. The original check can pass but the 
 broadcast_merge( result ,  proposals_pshape[0], dim_num_images) would fail.

**Review by jane-intel — APPROVED** [review]
